### PR TITLE
Add coverage annotation and refactor addIceCandidate test

### PR DESCRIPTION
**Pull request by soareschen**
The code is slightly rearranged to reflect the order of the test steps, so the diff is slightly hard to read.

### CODE REVIEW

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/244878611)

# Firefox (nightly) #
Testing web-platform-tests at revision 92af93a38b7f73f5e9de61aa167d0348503cafbe
Using browser at version BuildID 20170612100241; SourceStamp 27cad9749cddf68e11fdd4e5d73dad84a8f8cf23
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>1 test ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/6292/webrtc/RTCPeerConnection-addIceCandidate.html">/webrtc/RTCPeerConnection-addIceCandidate.html</a></summary>


|                                                      Subtest                                                      | Results |                                                                                                                                              Messages                                                                                                                                             |
|-------------------------------------------------------------------------------------------------------------------|---------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
|                                                                                                                   | OK      |                                                                                                                                                                                                                                                                                                   |
| `Add null candidate should reject with TypeError`                                                                 | FAIL    | `assert_unreached: Should have rejected: undefined Reached unreachable code`                                                                                                                                                                                                                      |
| `Add ICE candidate before setting remote description should reject with InvalidStateError`                        | PASS    |                                                                                                                                                                                                                                                                                                   |
| `Add candidate when remote description is null should never resolve when pc is closed`                            | PASS    |                                                                                                                                                                                                                                                                                                   |
| `Add ICE candidate after setting remote description should succeed`                                               | PASS    |                                                                                                                                                                                                                                                                                                   |
| `Add ICE candidate with RTCIceCandidate should succeed`                                                           | PASS    |                                                                                                                                                                                                                                                                                                   |
| `Add candidate with only valid sdpMid should succeed`                                                             | FAIL    | `promise_test: Unhandled rejection with value: object "TypeError: Missing required 'candidate' member of RTCIceCandidateInit."`                                                                                                                                                                   |
| `Add candidate with only valid sdpMLineIndex should succeed`                                                      | FAIL    | `promise_test: Unhandled rejection with value: object "TypeError: Missing required 'candidate' member of RTCIceCandidateInit."`                                                                                                                                                                   |
| `addIceCandidate with first sdpMid and sdpMLineIndex add candidate to first media stream`                         | PASS    |                                                                                                                                                                                                                                                                                                   |
| `addIceCandidate with second sdpMid and sdpMLineIndex should add candidate to second media stream`                | PASS    |                                                                                                                                                                                                                                                                                                   |
| `Add candidate for first media stream with null ufrag should add candidate to first media stream`                 | PASS    |                                                                                                                                                                                                                                                                                                   |
| `Adding multiple candidates should add candidates to their corresponding media stream`                            | PASS    |                                                                                                                                                                                                                                                                                                   |
| `Add with empty candidate string (end of candidate) should succeed`                                               | FAIL    | `promise_test: Unhandled rejection with value: object "InvalidCandidateError: Invalid candidate, no ':' ()"`                                                                                                                                                                                      |
| `Add candidate with both sdpMid and sdpMLineIndex manually set to null should reject with TypeError`              | PASS    |                                                                                                                                                                                                                                                                                                   |
| `Add candidate with only valid candidate string should reject with TypeError`                                     | PASS    |                                                                                                                                                                                                                                                                                                   |
| `Add candidate with invalid candidate string and both sdpMid and sdpMLineIndex null should reject with TypeError` | PASS    |                                                                                                                                                                                                                                                                                                   |
| `Add candidate with empty dict should reject with TypeError`                                                      | PASS    |                                                                                                                                                                                                                                                                                                   |
| `Add candidate with manually filled default values should reject with TypeError`                                  | PASS    |                                                                                                                                                                                                                                                                                                   |
| `Add candidate with invalid sdpMid should reject with OperationError`                                             | FAIL    | `assert_throws: function "function() { throw e }" threw object "InvalidCandidateError: Mismatch between mid and level - "invalid" is not the mid for level 0; "a1" is" that is not a DOMException OperationError: property "name" is equal to "InvalidCandidateError", expected "OperationError"` |
| `Add candidate with invalid sdpMLineIndex should reject with OperationError`                                      | FAIL    | `assert_throws: function "function() { throw e }" threw object "InvalidCandidateError: Index 2 out of range" that is not a DOMException OperationError: property "name" is equal to "InvalidCandidateError", expected "OperationError"`                                                           |
| `Invalid sdpMLineIndex should be ignored if valid sdpMid is provided`                                             | FAIL    | `promise_test: Unhandled rejection with value: object "InvalidCandidateError: Index 2 out of range"`                                                                                                                                                                                              |
| `Add candidate for media stream 2 with null ufrag should succeed`                                                 | PASS    |                                                                                                                                                                                                                                                                                                   |
| `Add candidate with invalid ufrag should reject with OperationError`                                              | FAIL    | `assert_unreached: Should have rejected: undefined Reached unreachable code`                                                                                                                                                                                                                      |
| `Add candidate with invalid candidate string should reject with OperationError`                                   | FAIL    | `assert_throws: function "function() { throw e }" threw object "InvalidCandidateError: Invalid candidate, no ':' ((Invalid) candidate \r\n string)" that is not a DOMException OperationError: property "name" is equal to "InvalidCandidateError", expected "OperationError"`                    |
| `Add candidate with sdpMid belonging to different ufrag should reject with OperationError`                        | FAIL    | `assert_unreached: Should have rejected: undefined Reached unreachable code`                                                                                                                                                                                                                      |
| `Add valid candidate should never resolve when pc is closed`                                                      | PASS    |                                                                                                                                                                                                                                                                                                   |
| `Add candidate with invalid candidate string should never resolve when pc is closed`                              | PASS    |                                                                                                                                                                                                                                                                                                   |
| `Add candidate with invalid sdpMid should never resolve when pc is closed`                                        | PASS    |                                                                                                                                                                                                                                                                                                   |
| `Add candidate with invalid sdpMLineIndex should never resolve when pc is closed`                                 | PASS    |                                                                                                                                                                                                                                                                                                   |

</details>
</details>

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/244878612)

# Sauce (safari) #
Testing web-platform-tests at revision 92af93a38b7f73f5e9de61aa167d0348503cafbe
Using browser at version 10.0
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>1 test ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/6292/webrtc/RTCPeerConnection-addIceCandidate.html">/webrtc/RTCPeerConnection-addIceCandidate.html</a></summary>


|                                                      Subtest                                                      | Results |                 Messages                 |
|-------------------------------------------------------------------------------------------------------------------|---------|------------------------------------------|
|                                                                                                                   | OK      |                                          |
| `Add null candidate should reject with TypeError`                                                                 | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add ICE candidate before setting remote description should reject with InvalidStateError`                        | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add candidate when remote description is null should never resolve when pc is closed`                            | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add ICE candidate after setting remote description should succeed`                                               | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add ICE candidate with RTCIceCandidate should succeed`                                                           | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add candidate with only valid sdpMid should succeed`                                                             | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add candidate with only valid sdpMLineIndex should succeed`                                                      | FAIL    | `Can't find variable: RTCPeerConnection` |
| `addIceCandidate with first sdpMid and sdpMLineIndex add candidate to first media stream`                         | FAIL    | `Can't find variable: RTCPeerConnection` |
| `addIceCandidate with second sdpMid and sdpMLineIndex should add candidate to second media stream`                | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add candidate for first media stream with null ufrag should add candidate to first media stream`                 | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Adding multiple candidates should add candidates to their corresponding media stream`                            | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add with empty candidate string (end of candidate) should succeed`                                               | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add candidate with both sdpMid and sdpMLineIndex manually set to null should reject with TypeError`              | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add candidate with only valid candidate string should reject with TypeError`                                     | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add candidate with invalid candidate string and both sdpMid and sdpMLineIndex null should reject with TypeError` | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add candidate with empty dict should reject with TypeError`                                                      | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add candidate with manually filled default values should reject with TypeError`                                  | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add candidate with invalid sdpMid should reject with OperationError`                                             | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add candidate with invalid sdpMLineIndex should reject with OperationError`                                      | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Invalid sdpMLineIndex should be ignored if valid sdpMid is provided`                                             | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add candidate for media stream 2 with null ufrag should succeed`                                                 | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add candidate with invalid ufrag should reject with OperationError`                                              | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add candidate with invalid candidate string should reject with OperationError`                                   | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add candidate with sdpMid belonging to different ufrag should reject with OperationError`                        | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add valid candidate should never resolve when pc is closed`                                                      | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add candidate with invalid candidate string should never resolve when pc is closed`                              | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add candidate with invalid sdpMid should never resolve when pc is closed`                                        | FAIL    | `Can't find variable: RTCPeerConnection` |
| `Add candidate with invalid sdpMLineIndex should never resolve when pc is closed`                                 | FAIL    | `Can't find variable: RTCPeerConnection` |

</details>
</details>

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/244878618)

# Chrome (unstable) #
Testing web-platform-tests at revision 92af93a38b7f73f5e9de61aa167d0348503cafbe

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/244878619)

# Sauce (MicrosoftEdge) #
Testing web-platform-tests at revision 92af93a38b7f73f5e9de61aa167d0348503cafbe

File: webrtc/RTCPeerConnection-addIceCandidate.html [hunk 1]
```diff
@@ -2,9 +2,39 @@
...
+        ...
+      };
+
+      dictionary RTCIceCandidateInit {
+        DOMString       candidate = "";
+        DOMString?      sdpMid = null;
+        unsigned short? sdpMLineIndex = null;
+        DOMString       ufrag;
+      };
+   */
```
Comment (rwaldron) [inline webrtc/RTCPeerConnection-addIceCandidate.html]: Excellent

**Comment by rwaldron** [comment]
I will do full review tomorrow AM

File: webrtc/RTCPeerConnection-addIceCandidate.html [hunk 2]
```diff
@@ -191,8 +300,19 @@
...
+  /*
+    4.3.2.  addIceCandidate
+      4.6.  If candidate.candidate is an empty string, process candidate as an
+            end-of-candidates indication for the corresponding media description
+            and ICE candidate generation.
+        2.  If candidate is applied successfully, the user agent MUST queue
+            a task that runs the following steps:
+          2.  Let remoteDescription be connection's pendingRemoteDescription
+              if not null, otherwise connection's currentRemoteDescription.
+          3.  Add candidate to remoteDescription.
```
Comment (rwaldron) [inline webrtc/RTCPeerConnection-addIceCandidate.html]: Marked with "untestable" annotations

**Review by rwaldron — APPROVED** [review]
Coverage spec updated: https://rwaldron.github.io/webrtc-pc/#dom-rtcpeerconnection-addicecandidate

**Pull request merged**